std-bits/base/src/main/java/org/enso/base/parser/NumberWithSeparators.java

```diff
@@ -0,0 +1,348 @@
+package org.enso.base.parser;
+
+import java.util.Arrays;
+import org.enso.base.parser.FormatDetectingNumberParser.NumberParseDouble;
+import org.enso.base.parser.FormatDetectingNumberParser.NumberParseFailure;
+import org.enso.base.parser.FormatDetectingNumberParser.NumberParseLong;
+import org.enso.base.parser.FormatDetectingNumberParser.NumberParseResult;
+
+/**
+ * Number parsing with separators. Two special cases: - Encounter a single . or , with 3 trailing
+ * numbers. - Could be either DOT_COMMA or COMMA_DOT. Hence, DOT_UNKNOWN and COMMA_UNKNOWN.
+ */
+public enum NumberWithSeparators {
+  UNKNOWN(Constants.UNKNOWN, Constants.UNKNOWN),
+
+  // Special case where we have encountered a . with 3 trailing digits. ##0.123 ####.123
+  DOT_UNKNOWN(Constants.UNKNOWN, '.'),
+  // Special case where we have encountered a single . within 3 digits from start and without 3
+  // digits from end.
+  UNKNOWN_DOT(Constants.UNKNOWN, '.'),
+  // Special case where we have encountered a , with 3 trailing digits.
+  COMMA_UNKNOWN(',', Constants.UNKNOWN),
+  // Special case where we have encountered a single . within 3 digits from start and without 3
+  // digits from end.
+  UNKNOWN_COMMA(Constants.UNKNOWN, '.'),
+
+  NO_UNKNOWN(Constants.NONE, Constants.UNKNOWN),
+  NO_DOT(Constants.NONE, '.'),
+  NO_COMMA(Constants.NONE, ','),
+
+  DOT_COMMA('.', ','),
+
+  COMMA_DOT(',', '.'),
+
+  SPACE_UNKNOWN(' ', Constants.UNKNOWN),
+  SPACE_DOT(' ', '.'),
+  SPACE_COMMA(' ', ','),
+
+  SWISS_UNKNOWN('\'', Constants.UNKNOWN),
+  SWISS_DOT('\'', '.'),
+  SWISS_COMMA('\'', ',');
+
+  public static NumberWithSeparators fromSeparators(String thousand, String decimal) {
+    if (thousand != null && thousand.length() > 1) {
+      throw new IllegalArgumentException("Invalid thousand separator (more than one character).");
+    }
+
+    if (decimal != null && decimal.length() > 1) {
+      throw new IllegalArgumentException("Invalid decimal separator (more than one character).");
+    }
+
+    char thousands =
+        thousand == null
+            ? Constants.UNKNOWN
+            : (thousand.isEmpty() ? Constants.NONE : thousand.charAt(0));
+    char decimals =
+        decimal == null
+            ? Constants.UNKNOWN
+            : (decimal.isEmpty() ? Constants.NONE : decimal.charAt(0));
+
+    // For these special cases if we know the decimal separator should fix it.
+    if (thousands == Constants.UNKNOWN) {
+      if (decimals == '.') {
+        return UNKNOWN_DOT;
+      } else if (decimals == ',') {
+        return UNKNOWN_COMMA;
+      }
+    }
+
+    var matched =
+        Arrays.stream(values())
+            .filter(v -> v.thousands == thousands && v.decimal == decimals)
+            .findAny();
+    if (matched.isEmpty()) {
+      throw new IllegalArgumentException("Invalid separators.");
+    }
```

Comment:
Given that we care about performance, I'm wondering if some kind of `switch` statement would be faster here. But I guess we need to run some benchmarks to see it in practice. So totally OK for now but just noting that I'd think of trying out some other approach.